File: exercism/scala/clock/src/main/scala/clock.scala
object Clock {
  def padRight(in: String, n: Int = 2): String = in.reverse.padTo(n, "0").mkString("").reverse

  def apply(h: Int, m: Int): Clock = {
    val (carry, min) =
      if (m < 0) m / 60 - 1 -> (m % 60 + 60)
      else m / 60 -> m % 60

    val hour = carry + h

    if (hour < 0) new Clock(hour % 24 + 24, min)
    else new Clock(hour % 24, min)
  }

  def apply(m: Int): Clock = Clock.apply(0, m)
}

class Clock(val h: Int, val m: Int) {
  import Clock._

  def +(o: Clock) = o match {
    case c: Clock => Clock(h + c.h, m + c.m)
  }

  def -(o: Clock) = o match {
    case c: Clock => Clock(h - c.h, m - c.m)
  }

  override def equals(o: Any) = o match {
    case c: Clock => c.h == h && c.m == m
    case _ => false
  }

  override def toString = {
    s"${padRight(h.toString)}:${padRight(m.toString)}"
  }

}

File: exercism/elixir/clock/clock.exs
defmodule Clock do
  defstruct hour: 0, minute: 0

  @spec py_rem(integer, integer) :: integer
  def py_rem(d, m) do
    r = rem(d, m)

    if r < 0 do
      r + m
    else
      r
    end
  end

  def py_div(d, m) do
    (d / m)
    |> Float.floor
    |> round
  end

  @spec to_minute(Clock) :: integer
  def to_minute(%Clock{hour: h, minute: m}) do
    h * 60 + m
  end

  @spec from_minute(integer) :: Clock
  def from_minute(m) do
    h = py_rem(py_div(m, 60), 24)
    min = py_rem(m, 60)

    %Clock { hour: h, minute: min }
  end

  @doc """
  Returns a string representation of a clock:

  iex> Clock.new(8, 9) |> to_string
  "08:09"
  """
  @spec new(integer, integer) :: Clock
  def new(hour, minute) do
    %Clock { hour: hour, minute: minute }
    |> to_minute
    |> from_minute
  end

  @doc """
  Adds two clock times:

  iex> Clock.add(10, 0) |> Clock.add(3) |> to_string
  "10:03"
  """
  @spec add(Clock, integer) :: Clock
  def add(%Clock{hour: hour, minute: minute}=c, add_minute) do
    min = c |> to_minute
    new_min = add_minute + min
    |> from_minute
  end

  defimpl String.Chars, for: Clock do
    def to_string(%Clock{hour: h, minute: m}) do
      if h < 10 do
        if m < 10 do
          "0#{h}:0#{m}"
        else
          "0#{h}:#{m}"
        end
      else
        if m < 10 do
          "#{h}:0#{m}"
        else
          "#{h}:#{m}"
        end
      end
    end
  end
end
